File: Linked List/C++/LC 61 - ROTATE LIST.cpp
// LEET CODE 61 - Rotate List
// QUESTION LINK - https://leetcode.com/problems/rotate-list/

class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (head == NULL) return NULL;

        ListNode* temp = head;
        int size = 0;

        // Count the size of the list
        while (temp != NULL) {
            size++;
            temp = temp->next;
        }

        k %= size;
        if (k == 0) return head;

        ListNode* slow = head;
        ListNode* fast = head;

        // Move fast pointer k steps ahead
        while (k--) {
            fast = fast->next;
        }

        // Move both pointers until fast reaches the end
        while (fast->next != NULL) {
            slow = slow->next;
            fast = fast->next;
        }

        // Perform the rotation
        fast->next = head;
        head = slow->next;
        slow->next = NULL;

        return head;
    }
};

File: Linked List/Java/LC 61 - ROTATE LIST.java
// LEET CODE 61 - Rotate List
// QUESTION LINK - https://leetcode.com/problems/rotate-list/

class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null) return null;

        ListNode temp = head;
        int size = 0;

        // Count the size of the list
        while (temp != null) {
            size++;
            temp = temp.next;
        }

        k %= size;
        if (k == 0) return head;

        ListNode slow = head;
        ListNode fast = head;

        // Move fast pointer k steps ahead
        while (k-- > 0) {
            fast = fast.next;
        }

        // Move both pointers until fast reaches the end
        while (fast.next != null) {
            slow = slow.next;
            fast = fast.next;
        }

        // Perform the rotation
        fast.next = head;
        head = slow.next;
        slow.next = null;

        return head;
    }
}
